Util/Tokenization/src/main/java/org/tribuo/util/tokens/impl/wordpiece/WordpieceTokenizer.java

```diff
@@ -1,31 +1,67 @@
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
 
 import org.tribuo.util.tokens.Token;
 import org.tribuo.util.tokens.Token.TokenType;
 import org.tribuo.util.tokens.Tokenizer;
+import org.tribuo.util.tokens.impl.WhitespaceTokenizer;
 
 import com.oracle.labs.mlrg.olcut.config.Config;
 import com.oracle.labs.mlrg.olcut.provenance.ConfiguredObjectProvenance;
 import com.oracle.labs.mlrg.olcut.provenance.impl.ConfiguredObjectProvenanceImpl;
 
+/**
+ * This Tokenizer is meant to be a reasonable approximation of the BertTokenizer
+ * defined <a href=
+ * "https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/tokenization_bert.py#L117">here</a>.
+ * Please see class definition for <code>BertTokenizer</code> (the line numbers
+ * may change.) Please see notes in WordpieceTokenizerTest for information about
+ * how we tested the similarity between this tokenizer and the referenced python
+ * implementation and for regression test examples that fail. In short, there
+ * are outstanding discrepancies for texts that include Arabic and other
+ * non-latin scripts that generate so many "[UNK]" tokens for an English-based
+ * BPE vocabulary as to render the discrepancies as practically meaningless.
+ * <p>
+ * As in the reference implementation, the input text is whitespace tokenized
+ * and each token is further tokenized to account for things like punctuation
+ * and Chinese characters. The resulting tokens are then applied to the
+ * wordpiece algorithm implemented in {@link Wordpiece} which is driven by an
+ * input vocabulary which matches tokens and token suffixes as it can. Any
+ * tokens that are not found in the input vocbulary are marked as "unknown".
+ */
 public class WordpieceTokenizer implements Tokenizer {
 
-    @Config
+    @Config(mandatory=true, description="an instance of Wordpiece which applies the 'wordpiece' algorithm")
     private Wordpiece wordpiece;
-    @Config
+    @Config(description="determines whether or not to lowercase the input text")
     private boolean toLowerCase = true;
-    @Config
-    private Tokenizer tokenizer = new WordpiecePreprocessTokenizer();
-    @Config
+    @Config(description="performs whitespace tokenization before 'basic' tokenizer is applied (see next)")
+    private Tokenizer whitespaceTokenizer = new WhitespaceTokenizer();
+    @Config(description="performs some tokenization work on the input text before the wordpiece algorithm is applied to each resulting token.")
+    private Tokenizer basicTokenizer = new WordpieceBasicTokenizer();
+    @Config(description="determins whether ornot to strip accents/diacritics from the input text")
     private boolean stripAccents = true;
-    @Config
+    @Config(description="a set of 'token' strings that should never be split regardless of whether they have e.g. punctuation in the middle.  No entries should have whitespace in them.")
```

Comment:
Steve will complain at me if you don't put a comma after "e.g.".

Reply:
Ha!  Ok.  Done.